web/auth/auth.go

```diff
@@ -288,6 +289,101 @@ func authorize(c echo.Context) error {
 	return c.Redirect(http.StatusFound, u.String()+"#")
 }
 
+type accessTokenReponse struct {
+	Type    string `json:"token_type"`
+	Scope   string `json:"scope"`
+	Access  string `json:"access_token"`
+	Refresh string `json:"refresh_token,omitempty"`
+}
+
+func accessToken(c echo.Context) error {
+	grant := c.FormValue("grant_type")
+	clientID := c.FormValue("client_id")
+	clientSecret := c.FormValue("client_secret")
+	instance := middlewares.GetInstance(c)
+
+	if grant == "" {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "the grant_type parameter is mandatory",
+		})
+	}
+	if clientID == "" {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "the client_id parameter is mandatory",
+		})
+	}
+	if clientSecret == "" {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "the client_secret parameter is mandatory",
+		})
+	}
+
+	client := &Client{}
+	if err := couchdb.GetDoc(instance, ClientDocType, clientID, client); err != nil {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "the client must be registered",
+		})
+	}
+	if subtle.ConstantTimeCompare([]byte(clientSecret), []byte(client.ClientSecret)) == 0 {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "invalid client_secret",
+		})
+	}
+
+	var err error
+	out := accessTokenReponse{
+		Type: "bearer",
+	}
+
+	switch grant {
+	case "authorization_code":
+		code := c.FormValue("code")
+		if code == "" {
+			return c.JSON(http.StatusBadRequest, echo.Map{
+				"error": "the code parameter is mandatory",
+			})
+		}
+		accessCode := &AccessCode{}
+		if err = couchdb.GetDoc(instance, AccessCodeDocType, code, accessCode); err != nil {
+			return c.JSON(http.StatusBadRequest, echo.Map{
+				"error": "invalid code",
+			})
+		}
+		out.Scope = accessCode.Scope
+		out.Refresh, err = client.CreateJWT(instance, RefreshTokenAudience, out.Scope)
+		if err != nil {
+			return c.JSON(http.StatusInternalServerError, echo.Map{
+				"error": "Can't generate refresh token",
+			})
+		}
+		// Delete the access code, it can be used only once
+		couchdb.DeleteDoc(instance, accessCode)
```

Comment:
It's not critical. It's just one trick between many to avoid attacks. But even if the deletion fails, it remains very difficult to operate this attack: an attacker still have to steal the authorization code (difficulty: medium) AND the client_secret (difficulty: very hard).